ipld/src/selector/walk.rs

```diff
@@ -0,0 +1,208 @@
+    where
+        F: Fn(&Progress<L>, &Ipld) -> Result<(), String> + Sync,
+        L: LinkResolver + Sync + Send + Clone,
+    {
+        self.walk_all(ipld, resolver, |prog, ipld, reason| -> Result<(), String> {
+            if let VisitReason::SelectionMatch = reason {
+                return callback(prog, ipld);
+            }
+            Ok(())
+        })
+        .await
+    }
+}
+
+/// Provides reason for callback in traversal for `walk_all`.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum VisitReason {
+    /// Ipld node visited was a specific match.
+    SelectionMatch,
+    /// Ipld node was visited while searching for matches.
+    SelectionCandidate,
+}
+
+#[async_trait]
+pub trait LinkResolver {
+    #[allow(unused_variables)]
+    /// Resolves a Cid link into it's respective Ipld node, if it exists.
+    async fn load_link(&self, link: &Cid) -> Result<Option<Ipld>, String> {
+        Err("load_link not implemented on the LinkResolver".into())
+    }
+}
+
+impl LinkResolver for () {}
+
+#[derive(Clone, Debug)]
+pub struct Progress<L = ()>
+where
+    L: Clone,
+{
+    resolver: Option<L>,
+    path: Path,
+}
+
+impl<L> Progress<L>
+where
+    L: LinkResolver + Sync + Send + Clone,
+{
+    /// Returns the path of the current progress
+    pub fn path(&self) -> &Path {
+        &self.path
+    }
+
+    #[async_recursion]
+    async fn walk_all<F>(
+        &mut self,
+        ipld: &Ipld,
+        selector: Selector,
+        callback: &F,
+    ) -> Result<(), Error>
+    where
+        F: Fn(&Progress<L>, &Ipld, VisitReason) -> Result<(), String> + Sync,
+    {
+        // Resolve any links transparently before traversing
+        if let Ipld::Link(cid) = ipld {
+            if let Some(resolver) = &self.resolver {
+                let mut node = resolver.load_link(cid).await.map_err(Error::Link)?;
+                while let Some(Ipld::Link(c)) = node {
+                    node = resolver.load_link(&c).await.map_err(Error::Link)?;
+                }
+
+                if let Some(n) = node {
+                    return self.walk_all(&n, selector, callback).await;
+                }
+            }
+
+            // Link did not resolve to anything, stop traversal
+            return Ok(());
+        }
+
+        let reason = if selector.decide() {
+            VisitReason::SelectionMatch
+        } else {
+            VisitReason::SelectionCandidate
+        };
+        callback(self, ipld, reason).map_err(Error::Custom)?;
+
+        // If Ipld is list or map, continue traversal, otherwise return
+        match ipld {
+            Ipld::Map(_) | Ipld::List(_) => (),
+            _ => return Ok(()),
+        }
+
+        match selector.interests() {
+            Some(interests) => {
+                for ps in interests {
+                    let v = match ipld.lookup_segment(&ps) {
+                        Some(ipld) => ipld,
+                        None => continue,
+                    };
+                    self.clone()
```

Comment:
I'm not really stoked about having to clone `self` here, I'm assuming it's to not mess up the path? My main gripe with this is that it puts a `Clone` bound on the link resolver which shouldn't really be necessary. What if we just call `traverse_node` on `self` here, and inside the `traverse_node` body we pop the path segment from the path again after the recursive call in order to restore it?

Reply:
yeah exactly, it's necessary, if you don't then the path will be modified and appended to the path. The reason this can't just be iteratively done is that the paths can branch off and explore recursively. In theory we could probably switch to not having any of this recursively done, but until it's benchmarked and we actually need the performance, it's not worth it

Reply:
I think doing it recursively is fine. calling something like `self.path.pop();` at the end of `traverse_node` always restores the path to its previous state and it would get rid of a bunch of `Clone` bounds, so the suggestion is not just for performance reasons but mainly from an API perspective

Reply:
Yeah actually great suggestion, I got too influenced from the go impl there and had assumed it was necessary.